### ISSUE TITLE
[BUG] Apache2 envvars File Parsing Issue

### ISSUE THREAD

**Issue by Mr-Mors** [WillyXJ/facileManager#462]
fM Version : **3.4.2**
fmDNS Version : **3.3.4**

**In raising this issue, I confirm the following:**

- [X] I have read and understood the [contributors guide](https://github.com/WillyXJ/facileManager/blob/master/.github/CONTRIBUTING.md).
- [X] I have checked that the bug-fix I am reporting can be replicated, or that the feature I am suggesting isn't already present.
- [X] I have checked that the issue I'm posting isn't already reported.
- [X] I have checked that the issue I'm posting isn't already solved and no duplicates exist in [closed issues](https://github.com/WillyXJ/facileManager/issues?q=is%3Aissue+is%3Aclosed) and [opened issues](https://github.com/WillyXJ/facileManager/issues)
- [X] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.

---

**Expected Behavior:**
Installing fmDNS should create a properly formed sudoers file.

**Actual Behavior:**
Created sudoers file looks like:
```
Defaults:www-data}
export APACHE_RUN_USER  !requiretty
Defaults:www-data}
export APACHE_RUN_USER  !env_reset
www-data}
export APACHE_RUN_USER  ALL=(root)      NOPASSWD: /usr/local/bin/php /usr/local/facileManager/fmDNS/client.php *
```
**Steps to reproduce:**
Fresh install with Apache/2.4.25 (Debian)

Issue appears to be that **getParameterValue** does not check if the result is multiple lines or valid. 
On my fresh docker container, the /etc/apache2/envvars file has **APACHE_RUN_USER** on two lines, see snippet of file below:
```
# Since there is no sane way to get the parsed apache2 config in scripts, some
# settings are defined via environment variables and then used in apache2ctl,
# /etc/init.d/apache2, /etc/logrotate.d/apache2, etc.
: ${APACHE_RUN_USER:=www-data}
export APACHE_RUN_USER
: ${APACHE_RUN_GROUP:=www-data}
export APACHE_RUN_GROUP
# temporary state file location. This might be changed to /run in Wheezy+1
: ${APACHE_PID_FILE:=/var/run/apache2$SUFFIX/apache2.pid}
export APACHE_PID_FILE
: ${APACHE_RUN_DIR:=/var/run/apache2$SUFFIX}
export APACHE_RUN_DIR
: ${APACHE_LOCK_DIR:=/var/lock/apache2$SUFFIX}
export APACHE_LOCK_DIR
# Only /var/log/apache2 is handled by /etc/logrotate.d/apache2.
: ${APACHE_LOG_DIR:=/var/log/apache2$SUFFIX}
export APACHE_LOG_DIR
```

**Recommend Resolution:**
Either just use the environment variable Apache exports or cleanup the output of getParameterValue.

**Comment by Mr-Mors** [comment]
Not necessarily the best way to fix the issue, but my fix.
This is a fix for the /usr/local/facileManager/functions.php

```
--- functions.php.orginal	2019-09-20 21:37:38.386616352 +0000
+++ functions.php	2019-09-20 21:30:48.735691158 +0000
@@ -930,7 +930,8 @@
 			$user = getParameterValue('^User', $httpdconf, ' ');
 			if ($user[0] == '$') {
 				$user_var = preg_replace(array('/\$/', '/{/', '/}/'), '', $user);
-				$user = getParameterValue($user_var, findFile('envvars'), '=');
+				$user_value = getParameterValue($user_var, findFile('envvars'), '=');
+				$user = preg_replace(array('/\$/', '/{/', '/}/', '/\n/', '/export.*$/'), '', $user_value); 
 			}
 			echo "User3: ". $user."\n:User3";
 			echo fM('  --> Detected ' . $web_server['app'] . " runs as '$user'\n");
```
[functions.php.patch.txt](https://github.com/WillyXJ/facileManager/files/3637835/functions.php.patch.txt)

**Comment by WillyXJ** [comment]
This is now implemented in [fM 3.5](http://www.facilemanager.com/download/) and later.

**Comment by Mr-Mors** [comment]
@WillyXJ 

Unfortunately the patch included with 3.5 still does not work. The following is important to make the sudoers file properly.

`, '/\n/', '/export.*$/'`

So the line should be:

`$user = preg_replace(array('/\$/', '/{/', '/}/', '/\n/', '/export.*$/'), '', getParameterValue($user_var, findFile('envvars'), '='));` 

Without it my sudoers file looks like the following on a fresh install:
```
Defaults:www-data
export APACHE_RUN_USER  !requiretty
Defaults:www-data
export APACHE_RUN_USER  !env_reset
www-data
export APACHE_RUN_USER  ALL=(root)      NOPASSWD: /usr/local/bin/php /usr/local/facileManager/fmDNS/client.php *
```

With the extra search and replace values my sudoers file looks like:
```
Defaults:www-data  !requiretty
Defaults:www-data  !env_reset
www-data        ALL=(root)      NOPASSWD: /usr/local/bin/php /usr/local/facileManager/fmDNS/client.php *
```

**Comment by WillyXJ** [comment]
Would you please post the relevant section from your Apache config? The fix seemed to work in my environment so I'd like to see what yours is like in order to come up with a better parsing solution.

**Comment by Mr-Mors** [comment]
Apache2 envvars file export section:
```
# Since there is no sane way to get the parsed apache2 config in scripts, some
# settings are defined via environment variables and then used in apache2ctl,
# /etc/init.d/apache2, /etc/logrotate.d/apache2, etc.
: ${APACHE_RUN_USER:=www-data}
export APACHE_RUN_USER
: ${APACHE_RUN_GROUP:=www-data}
export APACHE_RUN_GROUP
# temporary state file location. This might be changed to /run in Wheezy+1
: ${APACHE_PID_FILE:=/var/run/apache2$SUFFIX/apache2.pid}
export APACHE_PID_FILE
: ${APACHE_RUN_DIR:=/var/run/apache2$SUFFIX}
export APACHE_RUN_DIR
: ${APACHE_LOCK_DIR:=/var/lock/apache2$SUFFIX}
export APACHE_LOCK_DIR
# Only /var/log/apache2 is handled by /etc/logrotate.d/apache2.
: ${APACHE_LOG_DIR:=/var/log/apache2$SUFFIX}
export APACHE_LOG_DIR
```

**Comment by Mr-Mors** [comment]
Also, here is the version info:

root@ns01:/etc/apache2# apache2 -version
Server version: Apache/2.4.25 (Debian)
Server built:   2018-11-03T18:46:19

I will note that this is running in a Docker container, but Apache should be Apache. The container is built on the "php:5-apache" image.

**Comment by WillyXJ** [comment]
Since the first fix didn't work, would you be able to apply this commit to your client to ensure it works in your environment before I release it to the masses?

b4710b624d8522801cc3fe5650d5d008cd74c6c2

**Comment by WillyXJ** [comment]
This is now implemented in [fM 3.5.1](http://www.facilemanager.com/download/) and later.
